File: challenges/Technetium 2019/solution2.java
// you can also use imports, for example:
// import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

import java.util.*;

class Solution {
    public String solution(int[][] A) {
        // Implement your solution here
        Queue<Integer> q = new ArrayDeque<>();
        q.add(0);
        StringBuilder sb = new StringBuilder("" + A[0][0]);
        int m = A.length, n = A[0].length, p = m + n - 2;
        for (int s = 0; s < p;) {
            int v = -1;
            for (int t = q.size(); t > 0; --t) {
                int x = q.poll(), y = s - x;
                for (int i = 0; i < 2; ++i) {
                    int nx = x + i, ny = y + (i ^ 1);
                    if (nx >= 0 && nx < m && ny >= 0 && ny < n && A[nx][ny] > 0) {
                        q.add(nx);
                        v = Math.max(v, A[nx][ny]);
                        A[nx][ny] = -A[nx][ny];
                    }
                }
            }
            sb.append(v);
            ++s;
            for (int t = q.size(); t > 0; --t) {
                int x = q.poll(), y = s - x;
                if (A[x][y] == -v) {
                    q.add(x);
                }
            }
        }
        return sb.toString();
    }
}

File: challenges/Neon 2014/solution2.cpp
// you can use includes, for example:
// #include <algorithm>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(vector<int> &R, int X, int M) {
    // Implement your solution here
    const int n = R.size(), len = X << 1;
    if (M / len < n) {
        return -1;
    }
    vector<int> mini(n);
    for (int i = n - 1, center = len * n - X; i >= 0; --i, center -= len) {
        mini[i] = R[i] - center;
        if (i + 1 < n) {
            mini[i] = min(mini[i], mini[i + 1]);
        }
    }
    const int limit = M - len * n;
    int r = 0;
    for (int i = 0, center = X, move = 0; i < n; ++i, center += len) {
        const int d = R[i] - center;
        if (d > r) {
            move = max(move, min((d + mini[i]) / 2, limit));
        }
        r = max(r, abs(d - move));
    }
    return r;
}
